pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1798,96 +1809,114 @@ func validateAccessCredentials(field *k8sfield.Path, accessCredentials []v1.Acce
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueInvalid,
-					Message: fmt.Sprintf("%s must have exactly one propagationMethod set", field.Index(idx).String()),
-					Field:   field.Index(idx).Child("userPassword", "propagationMethod").String(),
-				})
-			}
-		}
+func validateAccessCredentialSSH(field *k8sfield.Path, sshPublicKey *v1.SSHPublicKeyAccessCredential, hasConfigDriveVolume bool) []metav1.StatusCause {
+	if sshPublicKey == nil {
+		return nil
+	}
 
-		if count != 1 {
+	var causes []metav1.StatusCause
+
+	sourceCount := 0
+	if sshPublicKey.Source.Secret != nil {
+		sourceCount++
+	}
+
+	methodCount := 0
+	if sshPublicKey.PropagationMethod.ConfigDrive != nil {
+		methodCount++
+		if !hasConfigDriveVolume {
 			causes = append(causes, metav1.StatusCause{
 				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: fmt.Sprintf("%s must have exactly one access credential type set", field.Index(idx).String()),
-				Field:   field.Index(idx).String(),
+				Message: fmt.Sprintf("%s requires a configDrive volume to exist when the configDrive propagationMethod is in use.", field.String()),
+				Field:   field.Child("sshPublicKey", "propagationMethod").String(),
 			})
+
 		}
+	}
+	if sshPublicKey.PropagationMethod.QemuGuestAgent != nil {
+		if len(sshPublicKey.PropagationMethod.QemuGuestAgent.Users) == 0 {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Message: fmt.Sprintf("%s requires at least one user to be present in the users list", field.String()),
+				Field:   field.Child("sshPublicKey", "propagationMethod", "qemuGuestAgent", "users").String(),
+			})
+		}
+		methodCount++
+	}
+
+	if sourceCount != 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s must have exactly one source set", field.String()),
+			Field:   field.Child("sshPublicKey", "source").String(),
+		})
+	}
+	if methodCount != 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s must have exactly one propagationMethod set", field.String()),
+			Field:   field.Child("sshPublicKey", "propagationMethod").String(),
+		})
+	}
 
+	return causes
+}
+
+func validateAccessCredentialUserPassword(field *k8sfield.Path, userPassword *v1.UserPasswordAccessCredential) []metav1.StatusCause {
+	if userPassword == nil {
+		return nil
+	}
+
+	sourceCount := 0
+	if userPassword.Source.Secret != nil {
+		sourceCount++
+	}
+
+	var causes []metav1.StatusCause
+	if sourceCount != 1 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("%s must have exactly one source set", field.String()),
+			Field:   field.Child("userPassword", "source").String(),
+		})
+	}
+
+	methodCount := 0
+	if userPassword.PropagationMethod.QemuGuestAgent != nil {
+		methodCount++
+	}
+	if methodCount != 1 {
```

Comment:
Same here. I think we can get rid of `methodCount`:

```suggestion
	if userPassword.PropagationMethod.QemuGuestAgent == nil {
```

Reply:
Done